exporters/metric/cortex/config_test.go

```diff
@@ -0,0 +1,255 @@
+var validYAML = []byte(`url: /api/prom/push
+remote_timeout: 30s
+push_interval: 5s
+name: Valid Config Example
+basic_auth:
+  username: user
+  password: password
+bearer_token: qwerty12345
+tls_config:
+  ca_file: cafile
+  cert_file: certfile
+  key_file: keyfile
+  server_name: server
+  insecure_skip_verify: true
+`)
+
+// YAML file with no remote_timout property. It should produce a Config struct without errors.
+var noTimeoutYAML = []byte(`url: /api/prom/push
+push_interval: 5s
+name: Valid Config Example
+basic_auth:
+  username: user
+  password: password
+bearer_token: qwerty12345
+tls_config:
+  ca_file: cafile
+  cert_file: certfile
+  key_file: keyfile
+  server_name: server
+  insecure_skip_verify: true
+`)
+
+// YAML file with both password and password_file properties. It should fail to produce a Config
+// struct since password and password_file are mutually exclusive.
+var noEndpointYAML = []byte(`remote_timeout: 30s
+name: Valid Config Example
+basic_auth:
+  username: user
+  password: password
+bearer_token: qwerty12345
+tls_config:
+  ca_file: cafile
+  cert_file: certfile
+  key_file: keyfile
+  server_name: server
+  insecure_skip_verify: true
+`)
+
+// YAML file with both bearer_token and bearer_token_file properties. It should fail to produce a
+// Config struct since bearer_token and bearer_token_file are mutually exclusive.
+var twoPasswordsYAML = []byte(`url: /api/prom/push
+remote_timeout: 30s
+name: Valid Config Example
+basic_auth:
+  username: user
+  password: password
+  password_file: passwordfile
+bearer_token: qwerty12345
+tls_config:
+  ca_file: cafile
+  cert_file: certfile
+  key_file: keyfile
+  server_name: server
+  insecure_skip_verify: true
+`)
+
+// configValidStruct is the resulting Config struct from reading validYAML.
+var twoBearerTokensYAML = []byte(`url: /api/prom/push
+remote_timeout: 30s
+name: Valid Config Example
+basic_auth:
+  username: user
+  password: password
+bearer_token: qwerty12345
+bearer_token_file: bearertokenfile
+tls_config:
+  ca_file: cafile
+  cert_file: certfile
+  key_file: keyfile
+  server_name: server
+  insecure_skip_verify: true
+`)
+
+// ValidConfig is the resulting Config struct from reading validYAML.
+var ValidConfig = Config{
+	Endpoint:      "/api/prom/push",
+	RemoteTimeout: "30s",
+	Name:          "Valid Config Example",
+	BasicAuth: map[string]string{
+		"username": "user",
+		"password": "password",
+	},
+	BearerToken:     "qwerty12345",
+	BearerTokenFile: "",
+	TLSConfig: map[string]string{
+		"ca_file":              "cafile",
+		"cert_file":            "certfile",
+		"key_file":             "keyfile",
+		"server_name":          "server",
+		"insecure_skip_verify": "1",
+	},
+	ProxyURL:     "",
+	PushInterval: "5s",
+	Client:       nil,
+}
+
+// initYAML creates a YAML file at a given filepath. It does not remove any created directories or
+// files.
+func initYAML(yamlBytes []byte, path string) error {
```

Comment:
I'd look into removing the file system dependency from your unit tests. [`Afero`](https://github.com/spf13/afero) is from the same folks who wrote `viper` and has good integration, so that you can abstract away the file system.

Reply:
I took a look at `Afero` and it seems that I would need to call `viper.SetFs(fs afero.Fs)` before I call `viper.ReadInConfig()`. Since Viper is only used in `config.go`, would you recommend adding an Option to provide an `afero.Fs` file system or keep using the os file system?

Reply:
Yeah, that'd probably be a clean way to do it. You can hide the `afero.Fs` dependency as an unexported field in the `Config` struct. `NewConfig` can default the value to `afero.NewOsFs()`, and create a new `WithFs(afero.Fs)` `Option`, so your tests can utilize an in memory file system, and pass it in to `NewConfig(... WithFs(inMemoryFS))` to override the defaulted `OsFs`.

Reply:
I've modified both files to use Afero and added the WithFileSystem Option. One downside I found is that `Afero` does not work well with relative paths -- Viper looks for the file at `"."`, which is the cortex package directory, but `Afero` views `"."` as something else. To get around this, I added the `WithFilepath(...)` to all the `NewConfig(...)` calls.